python_modules/dagster/dagster/_core/execution/calculate_data_time.py

```diff
@@ -0,0 +1,87 @@
+from typing import AbstractSet, Mapping, Optional, Tuple
+
+from ..definitions.events import AssetKey
+from ..instance import DagsterInstance
+
+ROOT_DATA_TAG = ".dagster/root_data_ids"
+
+
+def get_upstream_materialization_times_for_key(
+    instance: DagsterInstance,
+    asset_key: AssetKey,
+    upstream_asset_key_mapping: Mapping[str, AbstractSet[str]],
+) -> Mapping[str, Tuple[Optional[int], Optional[float]]]:
+    from dagster._core.events import DagsterEventType
+    from dagster._core.storage.event_log.base import EventRecordsFilter
+
+    def _get_root_keys(asset_key):
+        def _recursive(key_str):
+            upstream_key_strs = upstream_asset_key_mapping[key_str]
+            if not upstream_key_strs:
+                return {key_str}
+            return set().union(
+                *(_recursive(upstream_key_str) for upstream_key_str in upstream_key_strs)
+            )
+
+        return _recursive(asset_key.to_user_string())
+
+    def _get_root_data(record):
```

Comment:
This part was trickier for me to understand, especially because it also writes into storage as well. Would it be possible to add a comment describing what this function does?